### ISSUE TITLE
question: extracting set manipulation

### ISSUE THREAD

**Issue by douglasduteil** [systemjs/builder#418]
I'm interested in using the tracking and the set arithmetic out of the Systemjs Builder. 
It is something you would welcome ? 

My use case is mainly enhancing the bundle arithmetic to deal with more operations like 
- the "not" for patterns
- the complement operator 
- simple evaluation (C = A ∪ B)

So we will be able to do 
- the union of A and B, denoted A ∪ B
- the intersection of A and B, denoted A ∩ B.
- the difference of A and B, denoted A \ B (or A − B),
- the symmetric difference of A and B, denoted A ∆ B
- the Cartesian product of two sets A and B, denoted by A × B

Here is a draft:

```js
{
  'polyfills_bundle': 'index ∪ npm:babel-runtime',
  'common_bundle': '((src/pageA ∪ src/pageB) × src/common) / polyfills_bundle'
}
```

**Comment by douglasduteil** [comment]
Would be cool if you can give me some guidance about the current implementation so I can start working on it. I'm trying to find out : 

- What is the data model of nodes in the tree ?

```js
// TraceNode Basic Data Structure
// Inspired by https://github.com/systemjs/builder/blob/0.14.11/lib/trace.js#L285
{
  name: String,
  path: null,
  metadata: {
    build: Boolean || null,
    loader: ?,
    loaderModule: ?,
    timestamp: Date
  },
  deps: Array<String>,
  depMap: {[key: String]: String },
  source: String || null,
  fresh: Boolean,
  timestamp: Date,
  configHash: String,

  plugin: String || null,
  runtimePlugin: Boolean,

  deferredImports: null
}
```

- Can we treat the expressions with [jsep](http://jsep.from.so/) or [pegjs](http://pegjs.org/) ?
- Do you mind if I'm using the mathematical operators (`∪`, `×`, `∩`, `∆`) ?
- How do you solve circular dependencies ? You just remove it ?
- How do you detect and extract dependences cross formats ? Are you always transpiling ? Using regex over `System.register` ? Or walking the AST like [precinct](https://github.com/mrjoelkemp/node-precinct) ?

Can't wait to start hacking on it :+1:

**Comment by guybedford** [comment]
Thanks for looking into this! It would be really great to have a contribution here and supporting parenthesis has been a feature needed for a long time being tracked in https://github.com/systemjs/builder/issues/286.

Currently the operations we support are:

* Union: `A + B`
* Complement: `A - B`
* Intersection: `A & B`

The implementation is all in https://github.com/systemjs/builder/blob/master/lib/arithmetic.js. We have a sort of a "AST" represented in https://github.com/systemjs/builder/blob/master/lib/arithmetic.js#L14, which could certainly be upgraded to using an expression parser to introduce new notational features.

Once in this AST, we just read the operations left to right, applying the operations from https://github.com/systemjs/builder/blob/master/lib/arithmetic.js#L54, so that things like node type, formats and circular references should all be unnecessary to worry about here.

The current syntax is based on the operator with a single space character on each side of it specifically - `" + "`, we would need to maintain this for backwards compatibility.

Because this is core arithmetic code, PRs will have to be very carefully reviewed. Some points from a contribution perspective (at risk of sounding somewhat petty):

* The smaller the diff the better. Negative diffs are ideal! Reducing the maintenance burden by keeping the code size down is a major priority.
* The fewer new dependencies introduced (and in turn with the least dependencies themselves) the better. No new dependencies is also ideal, but I'd be open to using one of the parsers above.

Yes I'd be open to allowing the arithmetic operators to be aliased by the unicode operators of their set operations. I'm not sure I see the value in the cartesian product though. Also while the symmetric difference is completely valid in this context, I'm not sure it has any practical use case to merit its inclusion.

**Comment by guybedford** [comment]
The other syntax features we support are:

* Single module identifier `[A]` meaning just the module A and none of its dependencies
* Wildcard / glob paths - `A/*` or `A/**/*` this is expanded as part of a preprocessing step once we've created the "AST" into an "expanded AST" explicitly listing the modules. As such, it shouldn't affect this work, apart from from a tokenizing point of view.

**Comment by guybedford** [comment]
@douglasduteil did my comments put you off that much here!?

**Comment by douglasduteil** [comment]
Ha-ha yeah sorry.

Thanks for those precious informations @guybedford. That helps me a lot and I want to make a trial with bench to measure if I'm doing the right thing.

However I didn't took the time to make any progress in here. I got frustrated by the http2 performance tests I made back then https://github.com/geelen/jspm-server/issues/41 and I'm studying better solutions for testing "universal libs" like systemjs. https://github.com/douglasduteil/study-test-es6-universal-lib
